<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />

  <title>scirpy.datasets.maynard2020 &mdash; scirpy  documentation</title>

      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/scanpy.css" type="text/css" />
      <link rel="stylesheet" href="../_static/typehints.css" type="text/css" />

  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>


    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="scirpy.datasets.vdjdb" href="scirpy.datasets.vdjdb.html" />
    <link rel="prev" title="scirpy.datasets.wu2020_3k" href="scirpy.datasets.wu2020_3k.html" />
 


</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/scirpy_logo_bright.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                0.13.0rc2.dev5+gf6ef57e
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../data-structure.html">Data structure and usage principles</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.io">Input/Output: <code class="docutils literal notranslate"><span class="pre">io</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.pp">Preprocessing: <code class="docutils literal notranslate"><span class="pre">pp</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#get-get">Get: <code class="docutils literal notranslate"><span class="pre">get</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.tl">Tools: <code class="docutils literal notranslate"><span class="pre">tl</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.pl">Plotting: <code class="docutils literal notranslate"><span class="pre">pl</span></code></a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../api.html#module-scirpy.datasets">Datasets: <code class="docutils literal notranslate"><span class="pre">datasets</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../api.html#example-datasets">Example datasets</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="scirpy.datasets.wu2020.html">scirpy.datasets.wu2020</a></li>
<li class="toctree-l4"><a class="reference internal" href="scirpy.datasets.wu2020_3k.html">scirpy.datasets.wu2020_3k</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">scirpy.datasets.maynard2020</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../api.html#reference-databases">Reference databases</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.util">Utility functions: <code class="docutils literal notranslate"><span class="pre">util</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.ir_dist">IR distance utilities: <code class="docutils literal notranslate"><span class="pre">ir_dist</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../ir-biology.html">Immune receptor (IR) model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_docs.html">Developer Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../versions.html">Other versions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zzz_bibliography.html">Bibliography</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">scirpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../api.html">API</a></li>
      <li class="breadcrumb-item active">scirpy.datasets.maynard2020</li>
      <li class="wy-breadcrumbs-aside">
              <!-- User defined GitHub URL -->
              <a href="https://github.com/scverse/scirpy/tree/master/scirpy/datasets/__init__.py#L105-L132" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="scirpy-datasets-maynard2020">
<h1>scirpy.datasets.maynard2020<a class="headerlink" href="#scirpy-datasets-maynard2020" title="Permalink to this headline"></a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="scirpy.datasets.maynard2020">
<span class="sig-prename descclassname"><span class="pre">scirpy.datasets.</span></span><span class="sig-name descname"><span class="pre">maynard2020</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#scirpy.datasets.maynard2020" title="Permalink to this definition"></a></dt>
<dd><p>Return the dataset from <span id="id1">[<a class="reference internal" href="../zzz_bibliography.html#id2" title="Ashley Maynard, Caroline E. McCoach, Julia K. Rotow, Lincoln Harris, Franziska Haderk, D. Lucas Kerr, Elizabeth A. Yu, Erin L. Schenk, Weilun Tan, Alexander Zee, Michelle Tan, Philippe Gui, Tasha Lea, Wei Wu, Anatoly Urisman, Kirk Jones, Rene Sit, Pallav K. Kolli, Eric Seeley, Yaron Gesthalter, Daniel D. Le, Kevin A. Yamauchi, David M. Naeger, Sourav Bandyopadhyay, Khyati Shah, Lauren Cech, Nicholas J. Thomas, Anshal Gupta, Mayra Gonzalez, Hien Do, Lisa Tan, Bianca Bacaltos, Rafael Gomez-Sjoberg, Matthew Gubens, Thierry Jahan, Johannes R. Kratz, David Jablons, Norma Neff, Robert C. Doebele, Jonathan Weissman, Collin M. Blakely, Spyros Darmanis, and Trever G. Bivona. Therapy-induced evolution of human lung cancer revealed by single-cell RNA sequencing. Cell, 182(5):1232–1251.e22, September 2020. URL: https://doi.org/10.1016/j.cell.2020.07.017, doi:10.1016/j.cell.2020.07.017.">MMR+20</a>]</span> as AnnData object.</p>
<p>21k cells from NSCLC profiled with Smart-seq2, of which 3,500 have <a class="reference internal" href="../glossary.html#term-TCR"><span class="xref std std-term">TCRs</span></a>
and 1,500 have <a class="reference internal" href="../glossary.html#term-BCR"><span class="xref std std-term">BCRs</span></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Scirpy example datasets are managed through <a class="reference external" href="https://github.com/fatiando/pooch">Pooch</a>.</p>
<p>By default, the dataset will be downloaded into your operating system’s default
cache directory (See <a class="reference external" href="https://www.fatiando.org/pooch/latest/api/generated/pooch.os_cache.html#pooch.os_cache" title="(in Pooch vv1.7.0)"><code class="xref py py-func docutils literal notranslate"><span class="pre">pooch.os_cache()</span></code></a> for more details). If it has already been
downloaded, it will be retrieved from the cache.</p>
<p>You can override the default cache dir by setting the <code class="docutils literal notranslate"><span class="pre">SCIRPY_DATA_DIR</span></code> environment variable
to a path of your preference.</p>
</div>
<p>The raw FASTQ files have been obtained from <a class="reference external" href="https://www.ebi.ac.uk/ena/browser/view/PRJNA591860">PRJNA591860</a>
and processed using the nf-core <a class="reference external" href="https://github.com/nf-core/rnaseq">RNA-seq pipeline</a> to obtain
gene expression and TraCeR/BraCeR to reconstruct receptors.</p>
<p>The processed files have been imported and transformed into an <a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.9.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">anndata.AnnData</span></code></a>
object using the following script:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># ---</span>
<span class="c1"># jupyter:</span>
<span class="c1">#   jupytext:</span>
<span class="c1">#     cell_metadata_filter: endofcell,-all</span>
<span class="c1">#     formats: py:light,ipynb</span>
<span class="c1">#     notebook_metadata_filter: -kernelspec</span>
<span class="c1">#     text_representation:</span>
<span class="c1">#       extension: .py</span>
<span class="c1">#       format_name: light</span>
<span class="c1">#       format_version: &#39;1.5&#39;</span>
<span class="c1">#       jupytext_version: 1.14.4</span>
<span class="c1"># ---</span>

<span class="c1"># %load_ext autoreload</span>
<span class="c1"># %autoreload 2</span>

<span class="c1"># +</span>
<span class="kn">from</span> <span class="nn">glob</span> <span class="kn">import</span> <span class="n">glob</span>
<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="kn">import</span> <span class="n">Pool</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">pandas.testing</span> <span class="k">as</span> <span class="nn">pdt</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">scipy.sparse</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="kn">from</span> <span class="nn">mudata</span> <span class="kn">import</span> <span class="n">MuData</span>

<span class="kn">import</span> <span class="nn">scirpy</span> <span class="k">as</span> <span class="nn">ir</span>

<span class="n">DATASET_DIR</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;/data/datasets/Maynard_Bivona_2020_NSCLC/&quot;</span><span class="p">)</span>
<span class="c1"># -</span>

<span class="c1"># The dataset has been downloaded from ENA and then processed using BraCer, TraCeR and the nf-core RNA-seq pipeline</span>
<span class="c1"># using `salmon`.</span>
<span class="c1">#</span>
<span class="c1"># We previously processed this dataset with STAR + featureCounts, but following up the discussion</span>
<span class="c1"># on nf-core, featureCounts is not state-of-the art any more for estimating transcript abundances.</span>
<span class="c1"># We, therefore, switched to the nf-core RNA-seq pipeline and Salmon.</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span>
    <span class="s2">&quot;/data/genomes/hg38/annotation/gencode/gencode.v33.primary_assembly.annotation.gtf&quot;</span><span class="p">,</span>
    <span class="s2">&quot;r&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="k">as</span> <span class="n">gtf</span><span class="p">:</span>
    <span class="n">entries</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gtf</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
            <span class="k">continue</span>
        <span class="n">attrs</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">attrs</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">attrs</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)]</span>
        <span class="n">attrs</span> <span class="o">=</span> <span class="p">[(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">attrs</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">]</span>
        <span class="n">attrs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">attrs</span><span class="p">)</span>
        <span class="n">entries</span><span class="p">[</span><span class="n">attrs</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">attrs</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]</span>


<span class="n">ensg2symbol</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">entries</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s2">&quot;index&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">])</span>
<span class="c1"># remove PAR_ genes and duplicated symbols (~80)</span>
<span class="n">ensg2symbol</span> <span class="o">=</span> <span class="n">ensg2symbol</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">ensg2symbol</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;PAR_&quot;</span><span class="p">),</span> <span class="p">:]</span>
<span class="n">ensg2symbol</span> <span class="o">=</span> <span class="n">ensg2symbol</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">ensg2symbol</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(),</span> <span class="p">:]</span>
<span class="n">ensg2symbol</span>

<span class="n">sample_paths</span> <span class="o">=</span> <span class="n">glob</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">DATASET_DIR</span> <span class="o">/</span> <span class="s2">&quot;rnaseq_pipeline/salmon/SRR*&quot;</span><span class="p">))</span>

<span class="nb">len</span><span class="p">(</span><span class="n">sample_paths</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">read_salmon</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;quant type can be one of &quot;tpm&quot;, &quot;count&quot;, &quot;count_scaled&quot; &quot;&quot;&quot;</span>
    <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;quant.genes.sf&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ensg2symbol</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="n">res</span><span class="p">[</span><span class="s2">&quot;sample_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">name</span>
    <span class="n">res</span><span class="p">[</span><span class="s2">&quot;var&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;index&quot;</span><span class="p">:</span> <span class="s2">&quot;ensg&quot;</span><span class="p">})</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s2">&quot;ensg&quot;</span><span class="p">,</span> <span class="s2">&quot;symbol&quot;</span><span class="p">]]</span>
    <span class="p">)</span>
    <span class="n">res</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">csc_matrix</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;NumReads&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="n">res</span><span class="p">[</span><span class="s2">&quot;count_scaled&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">csc_matrix</span><span class="p">(</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;NumReads&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">/</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;EffectiveLength&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="p">)</span>
    <span class="n">res</span><span class="p">[</span><span class="s2">&quot;tpm&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">csc_matrix</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;TPM&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">res</span>


<span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="mi">16</span><span class="p">)</span> <span class="k">as</span> <span class="n">p</span><span class="p">:</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">read_salmon</span><span class="p">,</span> <span class="n">sample_paths</span><span class="p">[:],</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>

<span class="c1"># check that gene symbols are the same in all arrays</span>
<span class="n">pdt</span><span class="o">.</span><span class="n">assert_frame_equal</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;var&quot;</span><span class="p">],</span> <span class="n">res</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="s2">&quot;var&quot;</span><span class="p">])</span>

<span class="n">sample_ids</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;sample_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">res</span><span class="p">])</span>
<span class="n">count_mat</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">res</span><span class="p">])</span><span class="o">.</span><span class="n">tocsr</span><span class="p">()</span>
<span class="n">tpm_mat</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;tpm&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">res</span><span class="p">])</span><span class="o">.</span><span class="n">tocsr</span><span class="p">()</span>
<span class="n">count_mat_scaled</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;count_scaled&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">res</span><span class="p">])</span><span class="o">.</span><span class="n">tocsr</span><span class="p">()</span>

<span class="c1"># ## Read and sanitize metadata</span>

<span class="c1"># + endofcell=&quot;--&quot;</span>
<span class="c1"># # +</span>
<span class="n">sample_info</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="n">DATASET_DIR</span> <span class="o">/</span> <span class="s2">&quot;scripts/make_h5ad&quot;</span> <span class="o">/</span> <span class="s2">&quot;sra_sample_info.csv&quot;</span><span class="p">,</span> <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="n">cell_metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="n">DATASET_DIR</span> <span class="o">/</span> <span class="s2">&quot;scripts/make_h5ad&quot;</span> <span class="o">/</span> <span class="s2">&quot;cell_metadata.csv&quot;</span><span class="p">,</span>
    <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># combine metadata</span>
<span class="n">meta</span> <span class="o">=</span> <span class="n">sample_info</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
    <span class="n">cell_metadata</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s2">&quot;cell_ID&quot;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;cell_id&quot;</span>
<span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;Run&quot;</span><span class="p">)</span>
<span class="c1"># -</span>

<span class="n">meta</span> <span class="o">=</span> <span class="n">meta</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="s2">&quot;Assay Type&quot;</span><span class="p">,</span>
        <span class="s2">&quot;AvgSpotLen&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SRA Study&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ReleaseDate&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Bases&quot;</span><span class="p">,</span>
        <span class="s2">&quot;disease&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Biomaterial_provider&quot;</span><span class="p">,</span>
        <span class="s2">&quot;BioProject&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Isolate&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Sample Name&quot;</span><span class="p">,</span>
        <span class="s2">&quot;BioSample&quot;</span><span class="p">,</span>
        <span class="s2">&quot;BioSampleModel&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Bytes&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Center Name&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Consent&quot;</span><span class="p">,</span>
        <span class="s2">&quot;DATASTORE filetype&quot;</span><span class="p">,</span>
        <span class="s2">&quot;DATASTORE provider&quot;</span><span class="p">,</span>
        <span class="s2">&quot;DATASTORE region&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Experiment&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Instrument&quot;</span><span class="p">,</span>
        <span class="s2">&quot;LibraryLayout&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Library Name&quot;</span><span class="p">,</span>
        <span class="s2">&quot;LibrarySelection&quot;</span><span class="p">,</span>
        <span class="s2">&quot;cell_ID&quot;</span><span class="p">,</span>
        <span class="s2">&quot;LibrarySource&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Organism&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Platform&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gender&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SAMPLE_TYPE&quot;</span><span class="p">,</span>
        <span class="s2">&quot;TISSUE&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;columns&quot;</span><span class="p">,</span>
<span class="p">)</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;Age&quot;</span><span class="p">:</span> <span class="s2">&quot;age&quot;</span><span class="p">,</span>
        <span class="s2">&quot;smokingHx&quot;</span><span class="p">:</span> <span class="s2">&quot;smoking_status&quot;</span><span class="p">,</span>
        <span class="s2">&quot;stage.at.dx&quot;</span><span class="p">:</span> <span class="s2">&quot;stage_at_diagnosis&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="n">axis</span><span class="o">=</span><span class="s2">&quot;columns&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">meta</span><span class="o">.</span><span class="n">tail</span><span class="p">()</span>
<span class="c1"># --</span>

<span class="n">meta</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
    <span class="n">columns</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;sample_name&quot;</span><span class="p">:</span> <span class="s2">&quot;sample&quot;</span><span class="p">,</span>
        <span class="s2">&quot;histolgy&quot;</span><span class="p">:</span> <span class="s2">&quot;condition&quot;</span><span class="p">,</span>
        <span class="s2">&quot;patient_id&quot;</span><span class="p">:</span> <span class="s2">&quot;patient&quot;</span><span class="p">,</span>
        <span class="s2">&quot;biopsy_site&quot;</span><span class="p">:</span> <span class="s2">&quot;tissue&quot;</span><span class="p">,</span>
        <span class="s2">&quot;primary_or_metastaic&quot;</span><span class="p">:</span> <span class="s2">&quot;origin&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">meta</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">{</span><span class="s2">&quot;Adenocarcinoma&quot;</span><span class="p">:</span> <span class="s2">&quot;LUAD&quot;</span><span class="p">,</span> <span class="s2">&quot;Squamous&quot;</span><span class="p">:</span> <span class="s2">&quot;LSCC&quot;</span><span class="p">}[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">meta</span><span class="p">[</span><span class="s2">&quot;condition&quot;</span><span class="p">]</span>
<span class="p">]</span>

<span class="n">meta</span><span class="p">[</span><span class="s2">&quot;tissue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;lymph_node&quot;</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;LN&quot;</span> <span class="k">else</span> <span class="n">x</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">meta</span><span class="p">[</span><span class="s2">&quot;tissue&quot;</span><span class="p">]]</span>

<span class="n">meta</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">meta</span><span class="p">[</span><span class="s2">&quot;origin&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">(),</span> <span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;patient&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

<span class="n">meta</span><span class="p">[</span><span class="s2">&quot;origin&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
    <span class="p">{</span><span class="s2">&quot;Primary&quot;</span><span class="p">:</span> <span class="s2">&quot;tumor_primary&quot;</span><span class="p">,</span> <span class="s2">&quot;Metastatic&quot;</span><span class="p">:</span> <span class="s2">&quot;tumor_metastasis&quot;</span><span class="p">}</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">meta</span><span class="p">[</span><span class="s2">&quot;origin&quot;</span><span class="p">]</span>
<span class="p">]</span>

<span class="n">meta</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;patient&quot;</span><span class="p">,</span> <span class="s2">&quot;tissue&quot;</span><span class="p">,</span> <span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="s2">&quot;origin&quot;</span><span class="p">]]</span>

<span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;patient&quot;</span><span class="p">,</span> <span class="s2">&quot;tissue&quot;</span><span class="p">,</span> <span class="s2">&quot;condition&quot;</span><span class="p">,</span> <span class="s2">&quot;origin&quot;</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">col</span><span class="p">,</span> <span class="n">meta</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

<span class="c1"># ## build adata object</span>

<span class="n">has_tx</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)</span>
<span class="n">has_meta</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">meta</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>

<span class="n">has_all</span> <span class="o">=</span> <span class="n">has_tx</span> <span class="o">&amp;</span> <span class="n">has_meta</span>

<span class="nb">len</span><span class="p">(</span><span class="n">has_tx</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">has_meta</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">has_all</span><span class="p">)</span>

<span class="n">sample_id_mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">has_all</span><span class="p">))</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span>
    <span class="n">var</span><span class="o">=</span><span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;var&quot;</span><span class="p">],</span>
    <span class="n">X</span><span class="o">=</span><span class="n">tpm_mat</span><span class="p">[</span><span class="n">sample_id_mask</span><span class="p">,</span> <span class="p">:],</span>
    <span class="n">obs</span><span class="o">=</span><span class="n">meta</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">sample_ids</span><span class="p">[</span><span class="n">sample_id_mask</span><span class="p">],</span> <span class="p">:],</span>
<span class="p">)</span>

<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">count_mat</span><span class="p">[</span><span class="n">sample_id_mask</span><span class="p">,</span> <span class="p">:]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts_length_scaled&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">count_mat_scaled</span><span class="p">[</span><span class="n">sample_id_mask</span><span class="p">,</span> <span class="p">:]</span>

<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;symbol&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># ## add IR information</span>

<span class="n">adata_tcr</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_tracer</span><span class="p">(</span><span class="n">DATASET_DIR</span> <span class="o">/</span> <span class="s2">&quot;smartseq2_pipeline/TraCeR&quot;</span><span class="p">)</span>

<span class="n">adata_bcr</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_bracer</span><span class="p">(</span>
    <span class="n">DATASET_DIR</span> <span class="o">/</span> <span class="s2">&quot;smartseq2_pipeline/BraCeR/filtered_BCR_summary/changeodb.tab&quot;</span>
<span class="p">)</span>

<span class="n">adata_airr</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">merge_airr</span><span class="p">(</span><span class="n">adata_tcr</span><span class="p">,</span> <span class="n">adata_bcr</span><span class="p">)</span>

<span class="n">mdata</span> <span class="o">=</span> <span class="n">MuData</span><span class="p">({</span><span class="s2">&quot;gex&quot;</span><span class="p">:</span> <span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;airr&quot;</span><span class="p">:</span> <span class="n">adata_airr</span><span class="p">})</span>

<span class="c1"># ## check that all is right</span>

<span class="n">mdata_vis</span> <span class="o">=</span> <span class="n">mdata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="n">adata</span><span class="o">.</span><span class="n">layers</span>

<span class="n">adata_vis</span> <span class="o">=</span> <span class="n">mdata_vis</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">4000</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;cell_ranger&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_vis</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;patient&quot;</span><span class="p">,</span> <span class="s2">&quot;origin&quot;</span><span class="p">,</span> <span class="s2">&quot;CD8A&quot;</span><span class="p">,</span> <span class="s2">&quot;CD14&quot;</span><span class="p">])</span>

<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">chain_qc</span><span class="p">(</span><span class="n">mdata_vis</span><span class="p">)</span>

<span class="n">mdata_vis</span><span class="o">.</span><span class="n">update_obs</span><span class="p">()</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">group_abundance</span><span class="p">(</span>
    <span class="n">mdata_vis</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;airr:receptor_subtype&quot;</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;gex:patient&quot;</span>
<span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">group_abundance</span><span class="p">(</span>
    <span class="n">mdata_vis</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;airr:chain_pairing&quot;</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;gex:patient&quot;</span>
<span class="p">)</span>

<span class="c1"># ## save MuData</span>

<span class="n">mdata</span><span class="o">.</span><span class="n">write_h5mu</span><span class="p">(</span><span class="s2">&quot;maynard2020.h5mu&quot;</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;lzf&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://muon.readthedocs.io/en/latest/api/generated/muon.MuData.html#muon.MuData" title="(in muon)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MuData</span></code></a></p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="scirpy.datasets.wu2020_3k.html" class="btn btn-neutral float-left" title="scirpy.datasets.wu2020_3k" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="scirpy.datasets.vdjdb.html" class="btn btn-neutral float-right" title="scirpy.datasets.vdjdb" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Gregor Sturm, Tamas Szabo..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>