<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />

  <title>Data structure and usage principles &mdash; scirpy  documentation</title>

      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/scanpy.css" type="text/css" />
      <link rel="stylesheet" href="_static/typehints.css" type="text/css" />

  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>


    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorials" href="tutorials.html" />
    <link rel="prev" title="Scirpy: A Scanpy extension for analyzing single-cell immune-cell receptor sequencing data" href="index.html" />
 


</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/scirpy_logo_bright.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                0.12.2.dev185+ge9f3a68
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Data structure and usage principles</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#using-scirpy">Using scirpy</a></li>
<li class="toctree-l2"><a class="reference internal" href="#storing-airr-rearrangement-data-in-anndata">Storing AIRR rearrangement data in AnnData</a></li>
<li class="toctree-l2"><a class="reference internal" href="#chain-indices">Chain indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="#accessing-airr-data">Accessing AIRR data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#working-with-multimodal-data">Working with multimodal data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#common-function-parameters">Common function parameters</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="ir-biology.html">Immune receptor (IR) model</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="developer_docs.html">Developer Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="versions.html">Other versions</a></li>
<li class="toctree-l1"><a class="reference internal" href="zzz_bibliography.html">Bibliography</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">scirpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Data structure and usage principles</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/scverse/scirpy/blob/master/docs/data-structure.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="data-structure-and-usage-principles">
<h1>Data structure and usage principles<a class="headerlink" href="#data-structure-and-usage-principles" title="Permalink to this headline"></a></h1>
<section id="using-scirpy">
<h2>Using scirpy<a class="headerlink" href="#using-scirpy" title="Permalink to this headline"></a></h2>
<p>Import scirpy as</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">scirpy</span> <span class="k">as</span> <span class="nn">ir</span>
</pre></div>
</div>
<p>As a scverse core package, scirpy adheres to the workflow principles
<a class="reference external" href="https://scanpy.readthedocs.io/en/stable/usage-principles.html">laid out by scanpy</a>.:</p>
<blockquote>
<div><ul class="simple">
<li><p>The <a class="reference internal" href="api.html#api"><span class="std std-ref">API</span></a> is divided into <em>preprocessing</em> (<code class="docutils literal notranslate"><span class="pre">pp</span></code>), <em>tools</em> (<code class="docutils literal notranslate"><span class="pre">tl</span></code>),
and <em>plotting</em> (<code class="docutils literal notranslate"><span class="pre">pl</span></code>).</p></li>
<li><p>All functions work on <a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.8.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> or <a class="reference external" href="https://mudata.readthedocs.io/en/latest/api/generated/mudata.MuData.html#mudata.MuData" title="(in mudata)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MuData</span></code></a> objects.</p></li>
<li><p>The <a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.8.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> instance is modified inplace, unless a function
is called with the keyword argument <code class="docutils literal notranslate"><span class="pre">inplace=False</span></code>.</p></li>
</ul>
</div></blockquote>
<p>We decided to handle a few minor points differently to Scanpy:</p>
<blockquote>
<div><ul class="simple">
<li><p>Plotting functions with inexpensive computations (e.g. <a class="reference internal" href="generated/scirpy.pl.clonal_expansion.html#scirpy.pl.clonal_expansion" title="scirpy.pl.clonal_expansion"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pl.clonal_expansion()</span></code></a>)
call the corresponding tool (<a class="reference internal" href="generated/scirpy.tl.clonal_expansion.html#scirpy.tl.clonal_expansion" title="scirpy.tl.clonal_expansion"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.clonal_expansion()</span></code></a>) on-the-fly and
don’t store the results in the <a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.8.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object.</p></li>
<li><p>All plotting functions, by default, return a <a class="reference external" href="https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.html#matplotlib.axes.Axes" title="(in Matplotlib v3.7.1)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Axes</span></code></a> object,
or a list of such.</p></li>
</ul>
</div></blockquote>
</section>
<section id="storing-airr-rearrangement-data-in-anndata">
<span id="data-structure"></span><h2>Storing AIRR rearrangement data in AnnData<a class="headerlink" href="#storing-airr-rearrangement-data-in-anndata" title="Permalink to this headline"></a></h2>
<p>For instructions how to load data into scirpy, see <a class="reference internal" href="tutorials/tutorial_io.html#importing-data"><span class="std std-ref">Loading adaptive immune receptor repertoire (AIRR)-sequencing data with Scirpy</span></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Scirpy’s data structure was fundamentally changed in version 0.13.0. While previously, immune receptor
data was expanded into columns in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>, they are now stored as <a class="reference internal" href="glossary.html#term-awkward-array"><span class="xref std std-term">awkward array</span></a> in <code class="docutils literal notranslate"><span class="pre">adata.obsm</span></code>.
Fore more details … # TODO</p>
</div>
<p>Scirpy combines the <a class="reference external" href="https://docs.airr-community.org/en/latest/datarep/rearrangements.html">AIRR Rearrangement standard</a>
for representing adaptive immune receptor repertoire data with scverse’s <a class="reference external" href="https://anndata.readthedocs.io/en/latest/">AnnData</a> data structure.</p>
<p><code class="docutils literal notranslate"><span class="pre">AnnData`</span></code> combines a gene expression matrix (<code class="docutils literal notranslate"><span class="pre">.X</span></code>), gene-level annotations (<code class="docutils literal notranslate"><span class="pre">.var</span></code>) and
cell-level annotations (<code class="docutils literal notranslate"><span class="pre">.obs</span></code>) into a single object. Additionally, matrices aligned to the cells can be stored in <code class="docutils literal notranslate"><span class="pre">.obsm</span></code>.</p>
<figure class="align-default">
<a class="reference internal image-reference" href="_images/anndata.svg"><img alt="_images/anndata.svg" src="_images/anndata.svg" width="350px" /></a>
</figure>
<p>The AIRR rearrangement standard defines a set of fields to describe single receptor chains. One cell can have
multiple receptor chains. This relationship is represented as an <a class="reference internal" href="glossary.html#term-awkward-array"><span class="xref std std-term">awkward array</span></a> stored in <code class="docutils literal notranslate"><span class="pre">adata.obsm[&quot;airr&quot;]</span></code>.</p>
<p>The first dimension of the array represents the cells and is aligned to the <code class="docutils literal notranslate"><span class="pre">obs</span></code> axis of the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object.
The second dimension represents the number of chains per cell and is of variable length. The third dimension
is a :ref:~akward.RecordType` and represents fields defined in the rearrangement standard.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># adata.obsm[&quot;airr&quot;]</span>
<span class="p">[</span>
    <span class="c1"># cell0: 2 chains</span>
    <span class="p">[{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="s2">&quot;TRA&quot;</span><span class="p">,</span> <span class="s2">&quot;junction_aa&quot;</span><span class="p">:</span> <span class="s2">&quot;CADASGT...&quot;</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="s2">&quot;TRB&quot;</span><span class="p">,</span> <span class="s2">&quot;junction_aa&quot;</span><span class="p">:</span> <span class="s2">&quot;CTFDD...&quot;</span><span class="p">}],</span>
    <span class="c1"># cell1: 1 chain</span>
    <span class="p">[{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="s2">&quot;IGH&quot;</span><span class="p">,</span> <span class="s2">&quot;junction_aa&quot;</span><span class="p">:</span> <span class="s2">&quot;CDGFFA...&quot;</span><span class="p">}],</span>
    <span class="c1"># cell2: 0 chains</span>
    <span class="p">[],</span>
<span class="p">]</span>
</pre></div>
</div>
<p>This allows to losslessly store a complete AIRR rearrangement table in AnnData. The purpose of scirpy’s <a class="reference internal" href="api.html#api-io"><span class="std std-ref">IO module</span></a>
is to create AnnData objects with the corresponding <code class="docutils literal notranslate"><span class="pre">obsm</span></code> entries. At this point, chains are neither filtered, nor separated by locus.
This allows any scverse ecosystem package working with AIRR data to adopt the datastructure and to reuse scirpy’s IO functions
if they use <a class="reference internal" href="ir-biology.html#receptor-model"><span class="std std-ref">scirpy’s receptor model</span></a> or not.</p>
</section>
<section id="chain-indices">
<span id="id1"></span><h2>Chain indices<a class="headerlink" href="#chain-indices" title="Permalink to this headline"></a></h2>
<p>The <a class="reference internal" href="ir-biology.html#receptor-model"><span class="std std-ref">scirpy receptor model</span></a> allows up to two pairs of chains per cell. This representation
requires separation of chains by <a class="reference internal" href="glossary.html#term-Chain-locus"><span class="xref std std-term">locus</span></a> into <a class="reference internal" href="glossary.html#term-V-D-J"><span class="xref std std-term">VJ</span></a> and <a class="reference internal" href="glossary.html#term-V-D-J"><span class="xref std std-term">VDJ</span></a> chains,
and (optionally) filtering non-productive chains.</p>
<p>The <a class="reference internal" href="generated/scirpy.pp.index_chains.html#scirpy.pp.index_chains" title="scirpy.pp.index_chains"><code class="xref py py-func docutils literal notranslate"><span class="pre">index_chains()</span></code></a> function serves for this purpose. It creates an additional <a class="reference internal" href="glossary.html#term-awkward-array"><span class="xref std std-term">awkward array</span></a>
in <code class="docutils literal notranslate"><span class="pre">adata.obsm</span></code> that has the following structure:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># adata.obsm[&quot;chain_indices&quot;]</span>
<span class="p">[</span>
    <span class="c1"># cell0:</span>
    <span class="c1">#   * 1 VJ chain which is at index 0 in `adata.obsm[&quot;airr&quot;][0]`</span>
    <span class="c1">#   * 1 VDJ chain which is at in dex 1 in `adata.obsm[&quot;airr&quot;][0]`</span>
    <span class="c1">#   * multichain = False, because the chains does not have more than 2 VJ or VDJ chains</span>
    <span class="p">{</span><span class="s2">&quot;VJ&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;VDJ&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;multichain&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">},</span> <span class="c1"># single pair</span>
    <span class="c1"># cell1:</span>
    <span class="c1">#   * primary VJ chain is at index 0 in `adata.obsm[&quot;airr&quot;][1]`</span>
    <span class="c1">#   * secondary VJ chain is at index 2 in `adata.obsm[&quot;airr&quot;][1]`</span>
    <span class="c1">#   * etc.</span>
    <span class="p">{</span><span class="s2">&quot;VJ&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;VDJ&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="s2">&quot;multichain&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">},</span> <span class="c1"># dual IR</span>
<span class="p">]</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">obsm[&quot;chain_indices&quot;]</span></code> array could easily be adapted to other receptor models. For instance,
a library working with spatial <a class="reference internal" href="glossary.html#term-TCR"><span class="xref std std-term">TCR</span></a> data where each entry in <code class="docutils literal notranslate"><span class="pre">obs</span></code> corresponds to a “spot” with multiple cells rather
than a single cell could have a list with an arbitrary number of indices for the <code class="docutils literal notranslate"><span class="pre">&quot;VJ&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;VDJ&quot;</span></code> entries, respectively.
By using a different function for chain indexing, it would also be very straightforward to support non-IMGT loci (e.g.
from other species).</p>
</section>
<section id="accessing-airr-data">
<span id="id2"></span><h2>Accessing AIRR data<a class="headerlink" href="#accessing-airr-data" title="Permalink to this headline"></a></h2>
<p>Any scirpy function accessing AIRR data uses these indices in <code class="docutils literal notranslate"><span class="pre">adata.obsm[&quot;chain_indices&quot;]</span></code> to subset the awkward array in
<code class="docutils literal notranslate"><span class="pre">adata.obsm[&quot;airr&quot;]</span></code>. To retreive AIRR data convenientely, we added the <a class="reference internal" href="generated/scirpy.get.airr.html#scirpy.get.airr" title="scirpy.get.airr"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.get.airr()</span></code></a> function. It allows
to specify one or multiple fields and chains and returns a <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series" title="(in pandas v2.0.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.Series</span></code></a> or <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v2.0.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a>, respectively:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># retrieve the &quot;locus&quot; field of the primary VJ chain for each cell</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">ir</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">airr</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;VJ_1&quot;</span><span class="p">)</span>
<span class="n">AAACCTGAGAGTGAGA</span><span class="o">-</span><span class="mi">1</span>     <span class="n">TRA</span>
<span class="n">AAACCTGAGGCATTGG</span><span class="o">-</span><span class="mi">1</span>     <span class="n">TRA</span>
<span class="n">AAACCTGCACCTCGTT</span><span class="o">-</span><span class="mi">1</span>    <span class="kc">None</span>
<span class="o">...</span>
</pre></div>
</div>
<p>By using the <a class="reference internal" href="generated/scirpy.get.airr_context.html#scirpy.get.airr_context" title="scirpy.get.airr_context"><code class="xref py py-func docutils literal notranslate"><span class="pre">airr_context()</span></code></a> context manager, fields can be temporarily added to the <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code>
and used, e.g. for plotting:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">ir</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">airr_context</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;VJ_1&quot;</span><span class="p">):</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;VJ_1_locus&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="working-with-multimodal-data">
<span id="multimodal-data"></span><h2>Working with multimodal data<a class="headerlink" href="#working-with-multimodal-data" title="Permalink to this headline"></a></h2>
<p>The recommended way of working with paired gene expression (GEX) and AIRR data is to use the
<a class="reference external" href="https://muon.readthedocs.io/en/latest/notebooks/quickstart_mudata.html">MuData</a> container. <code class="docutils literal notranslate"><span class="pre">MuData</span></code> manages
multiple <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> objects that share observations and/or features.</p>
<p>After reading in AIRR data with the scirpy <a class="reference internal" href="api.html#api-io"><span class="std std-ref">IO module</span></a> and gene expression data with scanpy,
they can be merged in a <code class="docutils literal notranslate"><span class="pre">MuData</span></code> object. For instance:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_airr</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_10x_vdj</span><span class="p">(</span><span class="s2">&quot;all_contig_annotations.json&quot;</span><span class="p">)</span>
<span class="n">adata_gex</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_10x_h5</span><span class="p">(</span><span class="s2">&quot;filtered_feature_bc_matrix.h5&quot;</span><span class="p">)</span>
<span class="n">mdata</span> <span class="o">=</span> <span class="n">MuData</span><span class="p">({</span><span class="s2">&quot;airr&quot;</span><span class="p">:</span> <span class="n">adata_airr</span><span class="p">,</span> <span class="s2">&quot;gex&quot;</span><span class="p">:</span> <span class="n">adata_gex</span><span class="p">})</span>
</pre></div>
</div>
<p>Scirpy functions can be applied directly to the MuData object. By default, it will retrieve AIRR data from the <code class="docutils literal notranslate"><span class="pre">&quot;airr&quot;</span></code>
modality.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">chain_qc</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
</pre></div>
</div>
<p>All functions updating <code class="docutils literal notranslate"><span class="pre">obs</span></code> inplace update both <code class="docutils literal notranslate"><span class="pre">mdata.obs[f&quot;airr:{key_added}&quot;]</span></code> and <code class="docutils literal notranslate"><span class="pre">mata.mod[&quot;airr&quot;].obs[key_added]</span></code>.
This means you usually do not need to call <code class="xref py py-meth docutils literal notranslate"><span class="pre">mdata.update()</span></code> after running a scirpy function.</p>
<p>Should you prefer to not use MuData, this is entirely possible. All scirpy functions work as well on a single
<code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object that contains gene expression data in <code class="docutils literal notranslate"><span class="pre">adata.X</span></code> and AIRR data in <code class="docutils literal notranslate"><span class="pre">adata.obsm[&quot;airr&quot;]</span></code>.
Here is one way how the AIRR data can be merged into an AnnData object that already contains gene expression data:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Map each cell barcode to its respective numeric index (assumes obs_names are unique)</span>
<span class="n">barcode2idx</span> <span class="o">=</span> <span class="p">{</span><span class="n">barcode</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">adata_airr</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)}</span>
<span class="c1"># Generate a slice for the awkward array that retrieves the corresponding row from `adata_airr` for each</span>
<span class="c1"># barcode in `adata_gex`. `-1` will generate all &quot;None&quot;s for barcodes that are not in `adata_airr`</span>
<span class="n">idx</span> <span class="o">=</span> <span class="p">[</span><span class="n">barcode2idx</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">barcode</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">barcode</span> <span class="ow">in</span> <span class="n">adata_gex</span><span class="o">.</span><span class="n">obs_names</span><span class="p">]</span>
<span class="n">adata_gex</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;airr&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_airr</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;airr&quot;</span><span class="p">][</span><span class="n">idx</span><span class="p">]</span>
</pre></div>
</div>
</section>
<section id="common-function-parameters">
<h2>Common function parameters<a class="headerlink" href="#common-function-parameters" title="Permalink to this headline"></a></h2>
<p>Wherever applicable, scirpy’s functions take the following arguments:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">airr_mod</span></code> specifies the slot in <code class="docutils literal notranslate"><span class="pre">MuData</span></code> that contains the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object with AIRR data. This parameter is
ignored when working with AnnData directly. Defaults to <code class="docutils literal notranslate"><span class="pre">&quot;airr&quot;</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">airr_key</span></code> specifies the slot in <code class="docutils literal notranslate"><span class="pre">AnnData.obsm</span></code> that contains the awkward array with AIRR data. Defaults to <code class="docutils literal notranslate"><span class="pre">&quot;airr&quot;</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">chain_idx_key</span></code> specifies the slot in <code class="docutils literal notranslate"><span class="pre">AnnData.obsm</span></code> thtat contains the chain indices. Defaults to <code class="docutils literal notranslate"><span class="pre">&quot;chain_indices&quot;</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">inplace</span></code> defines if a function stores its results back in the AnnData/MuData object or returns them.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">key_added</span></code> defines the key (e.g. in <code class="docutils literal notranslate"><span class="pre">.obs</span></code>) where a function’s result is stored if <code class="docutils literal notranslate"><span class="pre">inplace=True</span></code>.</p></li>
</ul>
</div></blockquote>
<p>The <a class="reference internal" href="generated/scirpy.util.DataHandler.html#scirpy.util.DataHandler" title="scirpy.util.DataHandler"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataHandler</span></code></a> class ensures that these parameters are handled consistently across functions.</p>
<p><strong>For most use cases you can stick to the default and do not need to modify these parameters.</strong></p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Scirpy: A Scanpy extension for analyzing single-cell immune-cell receptor sequencing data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorials.html" class="btn btn-neutral float-right" title="Tutorials" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Gregor Sturm, Tamas Szabo..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>